packages/components/tests/unit/flamegraph.spec.js

```diff
@@ -0,0 +1,28 @@
+import { digestEventArray } from '@/lib/flamegraph.js';
+import scenario from '@/stories/data/scenario.json';
+import { buildAppMap } from '@appland/models';
+
+const accumulateValue = (sum, { value }) => sum + value;
+
+const validateNode = (node) => {
+  // children
+  expect(node).toHaveProperty('children');
+  expect(node.children).toBeInstanceOf(Array);
+  node.children.forEach(validateNode);
+  // name //
+  expect(node).toHaveProperty('name');
+  expect(typeof node.name).toBe('string');
+  // value //
+  expect(node).toHaveProperty('value');
+  expect(typeof node.value).toBe('number');
+  expect(Number.isInteger(node.value)).toBe(true);
+  expect(node.value).toBeGreaterThanOrEqual(node.children.reduce(accumulateValue, 0));
+  expect(node.value);
+};
+
+describe('digestEventArray', () => {
+  it('produce valid flamegraph tree', () => {
+    const { events } = buildAppMap(scenario).normalize().build();
+    validateNode(digestEventArray(events));
+  });
```

Comment:
This doesn't assert the flamegraph is valid, just that an exception is not raised. The data within `@/stories/data/scenario.json` should be stable enough to make assertions of specific values within the tree.

Reply:
0a8ef4a94c6a44a54cf0481ae781669948b2f56e